packages/cli/src/server/output.rs

```diff
@@ -1,157 +1,518 @@
+            let stderr = app.stderr.next_line();
+            tokio::select! {
+                Ok(Some(line)) = stdout => (app.result.platform, Some(line), None),
+                Ok(Some(line)) = stderr => (app.result.platform, None, Some(line)),
+                _ = futures_util::future::pending() => (app.result.platform, None, None),
+            }
+        });
+
+        tokio::select! {
+            new_line = join_all(next_stdout) => {
+                for (platform, stdout, stderr) in new_line {
+                    if let Some(stdout) = stdout {
+                        self.running_apps.get_mut(&platform).unwrap().stdout_line.push_str(&stdout);
+                        self.push_log(platform, BuildMessage {
+                            level: Level::INFO,
+                            message: MessageType::Text(stdout),
+                        })
+                    }
+                    if let Some(stderr) = stderr {
+                        self.running_apps.get_mut(&platform).unwrap().stderr_line.push_str(&stderr);
+                        self.push_log(platform, BuildMessage {
+                            level: Level::ERROR,
+                            message: MessageType::Text(stderr),
+                        })
+                    }
+                }
+            },
+            event = event => {
+                self.handle_input(event.unwrap().unwrap())?;
+            }
         }
+
+        return Ok(());
+    }
+
+    pub fn shutdown(&mut self) -> io::Result<()> {
+        // if we're a tty then we need to disable the raw mode
+        if self.is_tty {
+            disable_raw_mode()?;
+            stdout().execute(LeaveAlternateScreen)?;
+
+            // todo: print the build info here for the most recent build, and then the logs of the most recent build
+            for (platform, build) in self.build_logs.iter() {
+                if build.messages.is_empty() {
+                    println!("No build logs for {platform:?}");
+                    continue;
+                }
+                println!("Build logs for {platform:?}:");
+                for message in build.messages.iter() {
+                    match &message.message {
+                        MessageType::Cargo(t) => println!("{:?}", t),
+                        MessageType::Text(t) => println!("{}", t),
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn handle_input(&mut self, input: Event) -> io::Result<()> {
+        // handle ctrlc
+        if let Event::Key(key) = input {
+            if let KeyCode::Char('c') = key.code {
+                if key.modifiers.contains(KeyModifiers::CONTROL) {
+                    return Err(io::Error::new(io::ErrorKind::Interrupted, "Ctrl-C"));
+                }
+            }
+        }
+
+        if let Event::Key(key) = input {
+            if let KeyCode::Char('/') = key.code {
+                self.fly_modal_open = !self.fly_modal_open;
```

Comment:
May need to debounce user inputs like this or track key state over time.

As when I press/hold it to toggle visible then not visible straight away whilst I am pressing the key sometimes I see the modal flicker but rare to see if visible.

Perhaps there is a key down / key up event?
Or can add one with checking previous key state.
I'm running from RustRover.